### ISSUE TITLE
Style Guide

### ISSUE THREAD

**Issue by gdude2002** [python-discord/pydis-bot-core#2]
- [ ] Go over our current linting setup, and document it
- [ ] Review the documentation make sure we're happy with it - is it too strict? Wrong? Not strict enough?
- [ ] Create proper documentation that contributors can read up on and rely upon

**Comment by gdude2002** [comment]
Here's what we're currently using to lint. We should review this and decide on whether it's precisely what we need!

If an error code is **bolded** below, we have ignored it.

# Flake8

`Flake8` is a general-purpose linting tool that exists to be a fully-extensible linting tool, and it has a bunch of plugins to that end. They'll be given their own sections below. `Flake8` provides its own errors, and also includes the errors provided by `pycodestyle`.

**Line length limit**: 120 characters

## Flake8 Errors

Code | Example Message
--------|-------------------------
C901 | Code is too complex (based on complexity config)
... | ...
E999 | Failed to compile a file into an AST
... | ...
F401 | `module` imported but unused
F402 | import `module` from line `N` shadowed by loop variable
F403 | ‘from `module` import *’ used; unable to detect undefined names
F404 | future import(s) `name` after other statements
F405 | `name` may be undefined, or defined from star imports: `module`
F406 | ‘from `module` import *’ only allowed at module level
F407 | an undefined `__future__` feature name was imported
... | ...
F601 | dictionary key `name` repeated with different values
F602 | dictionary key variable `name` repeated with different values
F621 | too many expressions in an assignment with star-unpacking
F622 | two or more starred expressions in an assignment `(a, *b, *c = d)`
F631 | assertion test is a tuple, which are always `True`
... | ...
F701 | a `break` statement outside of a `while` or `for` loop
F702 | a `continue` statement outside of a `while` or `for` loop
F703 | a `continue` statement in a `finally` block in a loop
F704 | a `yield` or `yield from` statement outside of a function
F705 | a `return` statement with arguments inside a generator
F706 | a `return` statement outside of a function/method
F707 | an `except:` block as not the last exception handler
F721 F722 | doctest syntax error syntax error in forward type annotation
... | ...
F811 | redefinition of unused `name` from line `N`
F812 | list comprehension redefines `name` from line `N`
F821 | undefined name `name`
F822 | undefined name `name` in `__all__`
F823 | local variable `name` … referenced before assignment
F831 | duplicate argument `name` in function definition
F841 | local variable `name` is assigned to but never used
... | ...
F901 | `raise NotImplemented` should be raise `NotImplementedError`

## Pycodestyle Errors

Code | Example Message
--------|-------------------------
E101 | indentation contains mixed spaces and tabs
E111 | indentation is not a multiple of four
E112 | expected an indented block
E113 | unexpected indentation
E114 | indentation is not a multiple of four (comment)
E115 | expected an indented block (comment)
E116 | unexpected indentation (comment)
... | ...
E121 | continuation line under-indented for hanging indent
E122 | continuation line missing indentation or outdented
E123 | closing bracket does not match indentation of opening bracket’s line
E124 | closing bracket does not match visual indentation
E125 | continuation line with same indent as next logical line
E126 | continuation line over-indented for hanging indent
E127 | continuation line over-indented for visual indent
E128 | continuation line under-indented for visual indent
E129 | visually indented line with same indent as next logical line
E131 | continuation line unaligned for hanging indent
E133 | closing bracket is missing indentation
... | ...
E201 | whitespace after ‘(‘
E202 | whitespace before ‘)’
E203 | whitespace before ‘:’
E211 | whitespace before ‘(‘
... | ...
E221 | multiple spaces before operator
E222 | multiple spaces after operator
E223 | tab before operator
E224 | tab after operator
E225 | missing whitespace around operator
**E226** | missing whitespace around arithmetic operator
E227 | missing whitespace around bitwise or shift operator
E228 | missing whitespace around modulo operator
... | ...
E231 | missing whitespace after ‘,’, ‘;’, or ‘:’
... | ...
E241 | multiple spaces after ‘,’
E242 | tab after ‘,’
... | ...
E251 | unexpected spaces around keyword / parameter equals
... | ...
E261 | at least two spaces before inline comment
E262 | inline comment should start with ‘# ‘
E265 | block comment should start with ‘# ‘
E266 | too many leading ‘#’ for block comment
... | ...
E271 | multiple spaces after keyword
E272 | multiple spaces before keyword
E273 | tab after keyword
E274 | tab before keyword
E275 | missing whitespace after keyword
... | ...
E301 | expected 1 blank line, found 0
E302 | expected 2 blank lines, found 0
E303 | too many blank lines (3)
E304 | blank lines found after function decorator
E305 | expected 2 blank lines after end of function or class
E306 | expected 1 blank line before a nested definition
... | ...
E401 | multiple imports on one line
E402 | module level import not at top of file
... | ...
E501 | line too long (82 > 79 characters)
E502 | the backslash is redundant between brackets
... | ...
E701 | multiple statements on one line (colon)
E702 | multiple statements on one line (semicolon)
E703 | statement ends with a semicolon
E704 | multiple statements on one line (def)
E711 | comparison to None should be ‘if cond is None:’
E712 | comparison to True should be ‘if cond is True:’ or ‘if cond:’
E713 | test for membership should be ‘not in’
E714 | test for object identity should be ‘is not’
E721 | do not compare types, use ‘isinstance()’
E722 | do not use bare except, specify exception instead
E731 | do not assign a lambda expression, use a def
E741 | do not use variables named ‘l’, ‘O’, or ‘I’
E742 | do not define classes named ‘l’, ‘O’, or ‘I’
E743 | do not define functions named ‘l’, ‘O’, or ‘I’
... | ...
E901 | SyntaxError or IndentationError
E902 | IOError
... | ...
W191 | indentation contains tabs
... | ...
W291 | trailing whitespace
W292 | no newline at end of file
W293 | blank line contains whitespace
... | ...
W391 | blank line at end of file
... | ...
**W503** | line break before binary operator
W504 | line break after binary operator
W505 | doc line too long (82 > 79 characters)
... | ...
W601 | .has_key() is deprecated, use ‘in’
W602 | deprecated form of raising exception
W603 | ‘<>’ is deprecated, use ‘!=’
W604 | backticks are deprecated, use ‘repr()’
W605 | invalid escape sequence ‘x’
W606 | ‘async’ and ‘await’ are reserved keywords starting with Python 3.7

# Bugbear

This is a `Flake8` plugin which exists to find likely bugs and design problems in your code.

Note that the ignored `B9xx` codes are considered opinionated and are disabled by default - they must be explicitly enabled, along with a whitelist of every other code one wishes to use.

Code | Explanation
--------|-----------------
B001 | Do not use bare `except:`, it also catches unexpected events like memory errors, interrupts, system exit, and so on. Prefer `except Exception:`. If you're sure what you're doing, be explicit and write `except BaseException:`. Disable E722 to avoid duplicate warnings.
B002 | Python does not support the unary prefix increment. Writing `++n` is equivalent to `+(+(n))`, which equals `n`. You meant `n += 1`.
B003 | Assigning to `os.environ` doesn't clear the environment. Subprocesses are going to see outdated variables, in disagreement with the current process. Use `os.environ.clear()` or the `env=` argument to `Popen`.
B004 | Using `hasattr(x, '__call__')` to test if `x` is callable is unreliable. If `x` implements custom `__getattr__` or its `__call__` is itself not callable, you might get misleading results. Use `callable(x)` for consistent results.
B005 | Using `.strip()` with multi-character strings is misleading the reader. It looks like stripping a substring. Move your character set to a constant if this is deliberate. Use `.replace()` or regular expressions to remove string fragments.
B006 | Do not use mutable data structures for argument defaults. All calls reuse one instance of that data structure, persisting changes between them.
B007 | Loop control variable not used within the loop body. If this is intended, start the name with an underscore.
B008 | Do not perform calls in argument defaults. The call is performed only once at function definition time. All calls to your function will reuse the result of that definition-time call. If this is intended, assign the function call to a module-level variable and use that variable as a default value.
B009 | Do not call `getattr(x, 'attr')`, instead use normal property access: `x.attr`. Missing a default to `getattr` will cause an `AttributeError` to be raised for non-existent properties. There is no additional safety in using `getattr` if you know the attribute name ahead of time.
B010 | Do not call `setattr(x, 'attr', val)`, instead use normal property access: `x.attr = val`. There is no additional safety in using `setattr` if you know the attribute name ahead of time.
... | ...
B301 | Python 3 does not include `.iter*` methods on dictionaries. The default behavior is to return iterables. Simply remove the `iter` prefix from the method. For Python 2 compatibility, also prefer the Python 3 equivalent if you expect that the size of the dict to be small and bounded. The performance regression on Python 2 will be negligible and the code is going to be the clearest. Alternatively, use `six.iter*` or `future.utils.iter*`.
B302 | Python 3 does not include `.view*` methods on dictionaries. The default behavior is to return viewables. Simply remove the `view` prefix from the method. For Python 2 compatibility, also prefer the Python 3 equivalent if you expect that the size of the dict to be small and bounded. The performance regression on Python 2 will be negligible and the code is going to be the clearest. Alternatively, use `six.view*` or `future.utils.view*`.
B303 | The `__metaclass__` attribute on a class definition does nothing on Python 3. Use `class MyClass(BaseClass, metaclass=...)`. For Python 2 compatibility, use `six.add_metaclass`.
B304 | `sys.maxint` is not a thing on Python 3. Use `sys.maxsize`.
B305 | `.next()` is not a thing on Python 3. Use the `next()` builtin. For Python 2 compatibility, use `six.next()`.
B306 | `BaseException.message` has been deprecated as of Python 2.6 and is removed in Python 3. Use `str(e)` to access the user-readable message. Use `e.args` to access arguments passed to the exception.
... | ...
**B901** | Using `return x` in a generator function used to be syntactically invalid in Python 2. In Python 3 `return x` can be used in a generator as a return value in conjunction with `yield from`. Users coming from Python 2 may expect the old behavior which might lead to bugs. Use native `async def` coroutines or mark intentional `return x` usage with `# noqa` on the same line.
**B902** | Invalid first argument used for method. Use `self` for instance methods, and `cls` for class methods (which includes `__new__` and `__init_subclass__`) or instance methods of metaclasses (detected as classes directly inheriting from type).
**B903** | Use `collections.namedtuple` (or `typing.NamedTuple`) for data classes that only set attributes in an `__init__` method, and do nothing else. If the attributes should be mutable, define the attributes in `__slots__` to save per-instance memory and to prevent accidentally creating additional attributes on instances.
**B950** | Line too long. This is a pragmatic equivalent of pycodestyle's E501: it considers "max-line-length" but only triggers when the value has been exceeded by more than 10%. You will no longer be forced to reformat code due to the closing parenthesis being one character too far to satisfy the linter. At the same time, if you do significantly violate the line length, you will receive a message that states what the actual limit is. This is inspired by Raymond Hettinger's "Beyond PEP 8" talk and highway patrol not stopping you if you drive < 5mph too fast. Disable E501 to avoid duplicate warnings.

# Docstrings

This is a `Flake8` plugin which adds support for [pydocstyle](https://github.com/pycqa/pydocstyle) errors to the linter.

`Pydocstyle` does not check D203, D212, D213, D214, D215, D404, D405, D406, D407, D408, D409, D410, or D411 by default, and they must be enabled explicitly - by setting changing the `convention` option in `tox.ini`, or enabling each code explicitly.

The `D1xx` error codes only apply to public constructs, i.e. constructs with a public immediate parent, and without a single leading underscore in the name.

Code | Explanation
--------|-----------------
D100 | Missing docstring in public module
D101 | Missing docstring in public class
D102 | Missing docstring in public method
D103 | Missing docstring in public function
D104 | Missing docstring in public package
D105 | Missing docstring in magic method
D106 | Missing docstring in public nested class
D107 | Missing docstring in `__init__`
... | ...
D200 | One-line docstring should fit on one line with quotes
D201 | No blank lines allowed before function docstring
D202 | No blank lines allowed after function docstring
**D203** | 1 blank line required before class docstring
D204 | 1 blank line required after class docstring
D205 | 1 blank line required between summary line and description
D206 | Docstring should be indented with spaces, not tabs
D207 | Docstring is under-indented
D208 | Docstring is over-indented
D209 | Multi-line docstring closing quotes should be on a separate line
D210 | No whitespaces allowed surrounding docstring text
D211 | No blank lines allowed before class docstring
**D212** | Multi-line docstring summary should start at the first line
**D213** | Multi-line docstring summary should start at the second line
**D214** | Section is over-indented
**D215** | Section underline is over-indented
... | ...
D300 | Use """triple double quotes"
D301 | Use r""" if any backslashes in a docstring
D302 | Use u""" for Unicode docstrings
... | ...
D400 | First line should end with a period
D401 | First line should be in imperative mood
D401 | First line should be in imperative mood; try rephrasing
D402 | First line should not be the function’s “signature”
D403 | First word of the first line should be properly capitalized
**D404** | First word of the docstring should not be This
**D405** | Section name should be properly capitalized
**D406** | Section name should end with a newline
**D407** | Missing dashed underline after section
**D408** | Section underline should be in the line following the section’s name
**D409** | Section underline should match the length of its name
**D410** | Missing blank line after section
**D411** | Missing blank line before section
D412 | No blank lines allowed between a section header and its content
D413 | Missing blank line after last section
D414 | Section has no content

# Import Order

This `flake8` plugin simply checks that your imports are grouped and ordered correctly. Remember to set `application-import-names` and `application-package-names` for import names local to the project, and pydis package names respectively.

**Import order style**: pycharm

This plugin is only able to check module-scoped imports, and will ignore imports in conditionals or other scopes.

Code | Explanation
--------|-----------------
I100 | Your import statements are in the wrong order.
I101 | The names in your from import are in the wrong order.
I201 | Missing newline between import groups.
I202 | Additional newline in a group of imports.

# Quotes

This somewhat controvertial `flake8` plugin enforces a particular style of quotes when defining strings.

**Quote type**: Double (`"`)

This plugin adheres to [PEP8 conventions](https://www.python.org/dev/peps/pep-0008/#string-quotes) to avoid backslashes in strings. For example, the following is still valid:

```py
s = 'double "quotes" wrapped in singles are ignored'
```

Code | Example message
--------|--------------------------
Q0 | Replace bad quotes

# Tidy Imports

This `flake8` plugin exists to do exactly two things - complain about unnecessary import aliases, and complain about banned imports.

**Banned imports**: None.

Banned imports can be configured in the `tox.ini`, as shown:

```ini
[flake8]
banned-modules = mock = use unittest.mock!
                 urlparse = use six.moves.urllib.parse!
                 {python2to3}
```

Code | Example message
--------|--------------------------
I200 | Unnecessary import alias - rewrite as 'from foo import bar'.
I201 | Banned import `foo` used

# TODO

A `flake8` plugin that simply errors on lines that contain a TODO. It simply checks for lines that contain `TODO`, `FIXME`, or `XXX`.

Code | Example message
--------|--------------------------
T000 | Todo note found.

# Type Annotations

A `flake8` plugin that is used to "validate type annotations syntax as it was originally proposed by Guido van Rossum."

Code | Example message
--------|--------------------------
T800 | Missing spaces between parameter annotation and default value
T801 | Missing spaces in return type annotation

**Comment by hargoniX** [comment]
For the TODO error code maybe [flake8-todo](https://pypi.org/project/flake8-todo/) could be useful?

**Comment by jchristgit** [comment]
I think some of these plugins are really nice but all in all it's super strict and I'm not a fan at all.
Some concerns:

- `1 blank line required before class docstring` does this suggest the following:
```py
class Thing:

    """
    Hello
    """
```
or is this inside of the docstring? Either way, I think it's unnecessary


- The TODO plugin. There's often things that we can't fix right away and a `TODO` or `FIXME` is a good way to keep it in to update it later.
- The Quotes plugin. That's such a tiny thing and I don't think we should use a plugin to enforce it.

**Comment by gdude2002** [comment]
@hargoniX You mean this flake8-todo? https://github.com/python-discord/pydis-bot-core/blob/master/setup.py#L35

@jchristgit Yes it does, but we're ignoring `D203`

I think we're gonna fundamentally disagree on the quotes thing to be honest, but I agree with you on the TODO plugin. That said, they _should_ be warnings rather than errors, at least, that's what the codes imply.

**Comment by jchristgit** [comment]
> Yes it does, but we're ignoring D203

Ah, I'm bling.

> they should be warnings rather than errors, at least, that's what the codes imply.

A warning won't fail the linting then? That's alright then, I guess

I also think we should set the line length to 100, not 120. I rarely see code where I think 120 columns do more good than harm

**Comment by gdude2002** [comment]
I am kind of curious about the generally-accepted line length limit - maybe we should poll on that one, I've seen a bunch of people suggest 100.

**Comment by jos-b** [comment]
In our current codebase on site the only lines over 100 are comments, documentation and (not surprisingly) our immensely complex database queries. I'm in favor of a 100 character per line limit.

**Comment by GhostofGoes** [comment]
What are thoughts on requiring PEP-257 for docstrings? (This was discussed before in dev-contrib, but figured I'd bring it up again here...sorry lemon)

**Comment by MarkKoz** [comment]
We should ignore T800 and T801. flake8 can detect this without any plugins. The only thing flake8-type-annotations is good for now is the spaces around the return value type annotation.

**Comment by gdude2002** [comment]
> We should ignore T800. flake8 can detect this without any plugins. The only thing flake8-type-annotations is good for now is the spaces around the return value type annotation.

What error does Flake8 throw for that?

---

I think I agree with a line length of 100. I don't want to budge on the quote style as it's probably the main thing I see people talking about - and it really doesn't make sense for it to not be consistent. As @sco1 said, there's no real pain in doing it either.

Although I disagree that we should use `black`. I draw the line there. And besides, I hate the style that `black` forces on you anyway.

---

I'm not really sure about `T000` - I think they should be output to the build log at some point just so that we have them, but let's remove it if it fails the build.

---

I'd like some more input here, especially from the @python-discord/contributors team - this is an important subject!

**Comment by scragly** [comment]
I believe I was the one who mentioned yesterday about personally using single quotes for keys and doubles for the rest. This is a personal choice, but I brought up the subject because I was asking if we could have an established style guide so I didn't have to think "which way do they want it" or go scrolling through the majority of the existing project to attempt to grasp the currently established style, which was inconsistent anyhow since there was no guide.

What's being done here is _exactly_ what I was hoping to see. It's no significant effort to conform to a style that's documented in a guide, and even less so when there are linters setup especially for checking the style for you. I truly don't care about quote's one way or another, as consistency was my primary concern.

I agree that TODOs should not fail a build. They are important indicators for overarching works. Something may not have been in the scope of a PR being worked on, and it's a pain to break away from focus to write up an Issue immediately so it's not forgotten later. 

I don't think `black` is necessary.

I like the line length of 100. I am used to 79(72) as outlined in PEP008, so 100 gives me wiggle room without making the col length unreasonably long, impacting the workspace layout. I really don't ever want to side-scroll code. 120 was too long and often had cases of this. It also doesn't look or feel clean while reading, as it's way beyond the comfortable skim-reading length.

I would like it if the start line of multi-line docstring summaries were consistent also (`D212`/`D213`). First or second, I don't mind which, so long as it's consistent. In the case both are allowed but there's a rule for usage, that should be outlined at the least in the Guide clearly, but this has the downfall of not being checked on lint.

D401 is present above twice, mistake maybe?

> What error does Flake8 throw for that? (`T800` native)

`E252 missing whitespace around parameter equals`

**Comment by MarkKoz** [comment]
Line length of 100 is good. I find myself breaking up code even when I know it could fit into 120 lines.

---

Not sure about allowing TODOs. I use them a lot, but I think it would be more organised of us to write up issues for things. 

> it's a pain to break away from focus to write up an Issue immediately so it's not forgotten later.

I like to write TODOs and go through and write issues for them once a push/merge is ready.

---

Quotes are such a tiny thing; they should not be a big deal and we should not overcomplicate them. Let's be consistent and use one type of quote for _everything_. Please, no doubles for one thing and single for another. I don't care for single/double - I just want consistency.

---

I'd like to also see some discussion on docstring writing style rather than formatting. By this I mean, for example, using the imperative mood or not discussing implementation details in the summary.

**Comment by gdude2002** [comment]
> I'd like to also see some discussion on docstring writing style rather than formatting. By this I mean, for example, using the imperative mood or not discussing implementation details in the summary.

Those are covered by the docstrings linter - Imperative mood, and no method functions in docstrings

**Comment by GhostofGoes** [comment]
I prefer a line length of 100.

Let's avoid `black`. It goes against what were are doing here, which is defining a style that fits our projects.

While I tend to have the same sentiment as Scragly about TODOs, in this case I agree with Mark. TODOs represent *something* that probably should be done at some point, and therefore should be in an issue somewhere. When they're left in the code, they are often neglected until someone else comes along to that part of the code to do something else. That person won't necessarily know the context behind the TODO, and the person who created it may not be available or even remember why they themselves created it in the first place.

My personal preference is for single quotes. I tend to use single quotes for everything except user output, such as logging (and docstrings of course).

**Comment by gdude2002** [comment]
Okay, this is what I'm getting from all this. If there are no objections, I'm going to implement it. The changes from the original issue:

* Change the line length limit to `100` characters
* Ignore `T800` (it's a duplicate of a basic Flake8 code)
* Stick with double quotes
* The thread seems to be just slightly in favour of keeping the TODOs plugin

**Comment by scragly** [comment]
After thinking about it more, and after seeing Mark's response, I'm going to retract my argument that TODOs being helpful for marking sections outside of current work scope counting towards a reason to not fail them. 

It's true that we can still use them during development, but convert them to Issue tickets before pushing.

There's still a chance they have a place with in-code markers for large projects that involve multiple stages, but that can be logged with feature tickets and comments anyway.

Due to this, I can only imagine it would be a problem for those who push incomplete, mid-work commits to the main repo (gdude has said both master and feature branches require CI builds to pass). I'd personally prefer if that behaviour was avoided though to prevent useless commit spam in logs.
